Functional interface for handling SSM parameters

slinderman opened the pull request:
Per our meeting last week, let's move to a functional interface for handling SSM parameters. Parameters will be stored in a nested dict whose leaf values are arrays. We'll keep another nested dict with the same key structure but whose leaf values are `ParameterProperties` objects: a simple dataclass that specifies whether or not the parameter is `trainable` and a `constrainer` to map from unconstrained parameterization to the more common, constrained form. (The `constrainer` is a tensorflow probability bijector that can map i both directions.) 

The ultimate goal is for parameters to live outside the model. Then the model is simply a collection of functions and fixed hyper-parameters, much like other JAX libraries like Flax, GPJax, etc.

murphyk commented on ssm_jax/parameters.py:
```diff
@@ -0,0 +1,58 @@
+        params (dict): combined dictionary of unconstrained and fixed parameters
+            in their natural (constrained) form.
+    """
+    params = dict()
```
shouldn't you initialize `params` to `fixed_params` and then just add to that?

slinderman replied:
Ahh, good suggestion! Let me see if that works.

murphyk commented on ssm_jax/parameters.py:
```diff
@@ -0,0 +1,58 @@
+        else:
+            params[k] = param_props[k].constrainer(v)
+    for k, v in fixed_params.items():
+        if not isinstance(v, dict):
```
what if v is a dict?

slinderman replied:
That case is already handled by the unconstrained loop above. This just copies over any fixed params into the current dict. It’s not the most self evident code though….

slinderman replied:
simplified per your suggestion above!

murphyk commented on ssm_jax/parameters_test.py:
```diff
@@ -0,0 +1,23 @@
+import tensorflow_probability.substrates.jax.bijectors as tfb
+
+
+def test_parameter_tofrom_unconstrained():
```
Please also add a test for the nested dictionary case

slinderman commented on ssm_jax/abstractions.py:
```diff
@@ -217,22 +158,3 @@ def _loss_fn(params, minibatch_emissions, **minibatch_covariates):
         self.unconstrained_params = params
         return losses
```
Good call. I’m not quite done with this refactor yet FYI. Will mark as “ready for review” and ping you on slack when it is… this will be a pretty big change though!

slinderman commented:
@murphyk: I think this is finally ready to be merged. Like I said, it touches literally everything in the codebase, often with significant changes, so it might be impossible to properly review. Since I was going through every file, I used the opportunity to do some cleanup and standardization:
- I refactored the HMM tests to share a simple template
- I tried to standardize LinearGaussianSSM and HMM code, and move shared functions into the shared SSM abstraction
- I generally cleaned up the HMM notebooks, but there is still work to do!
- I added a `warnings.py` file that suppresses known TFP warnings. That way we don't have to do it in every single notebook.

I couldn't do everything though. I ended up commenting out the kmeans initializations for the GMM-HMMs. It will take more effort to make that work with our new parameter setup, but I'd rather get this merged first and track that with a separate issue.

slinderman commented:
On my machine, only. the following tests fail:
> FAILED ssm_jax/hmm/models/test_models.py::test_sample_and_fit[MultivariateNormalDiagHMM-kwargs8-covariates8] - assert DeviceArray(False, dtype=bool)
> FAILED ssm_jax/hmm/models/test_models.py::test_sample_and_fit[MultivariateNormalTiedHMM-kwargs10-covariates10] - assert DeviceArray(False, dtype=bool)
> FAILED ssm_jax/structural_time_series/sts_test.py::test_sts_fit_hmc - ModuleNotFoundError: No module named 'tensorflow'

I think the first two are real bugs in the M-steps. The third seems to be a new tensorflow dependency added by sts. Can this be avoided?

murphyk commented on ssm_jax/hmm/demos/poisson_hmm_earthquakes.py:
```diff
@@ -11,187 +11,176 @@
-    fig.show()
-
+# earthquake data from http://earthquake.usgs.gov/
+EARTHQUAKES = jnp.array([
```
we should put all the data on one line and turn of black line length / format checking using `# fmt: off`
